apps/src/applab/ai.js

```diff
@@ -8,63 +8,105 @@ function generateCodeDesignElements(modelId, modelData) {
-      select.options.remove(0);
-      select.options.remove(0);
-      Object.keys(modelData.featureNumberKey[feature]).forEach(option => {
-        var optionElement = document.createElement('option');
-        optionElement.text = option;
-        select.options.add(optionElement);
-      });
+  if (modelData.features && modelData.label) {
+    modelData.features.forEach(feature => {
       y = y + SPACER_PIXELS;
-    } else {
-      // create text input field for each continuous feature
-      label.textContent = feature;
-      var labelMinMax = designMode.createElement('LABEL', x, y);
-      // return a string of min and max values rounded to two decimal places
-      var min = modelData.extremumsByColumn[feature].min.toFixed(2);
-      var max = modelData.extremumsByColumn[feature].max.toFixed(2);
-      // unary plus operator returns a number and truncates trailing zeroes
-      labelMinMax.textContent = `(min: ${+min}, max: ${+max}):`;
-      labelMinMax.style.width = '300px';
+      var label = designMode.createElement('LABEL', x, y);
+      var alphaNumFeature = stripSpaceAndSpecial(feature.id);
+      let fieldId;
+      label.id = 'design_' + alphaNumFeature + '_label';
+      label.style.width = '300px';
       y = y + SPACER_PIXELS;
-      var input = designMode.createElement('TEXT_INPUT', x, y);
-      fieldId = alphaNumFeature + '_input';
-      input.id = 'design_' + fieldId;
+      if (feature.values) {
+        // Create dropdown menu for each categorical feature.
+        label.textContent = feature.id + ':';
+        fieldId = alphaNumFeature + '_dropdown';
+        var select = designMode.createElement('DROPDOWN', x, y);
+        select.id = 'design_' + fieldId;
+        // App Lab automatically adds "option 1" and "option 2", remove them.
+        select.options.remove(0);
+        select.options.remove(0);
+        feature.values.forEach(option => {
+          var optionElement = document.createElement('option');
+          optionElement.text = option;
+          select.options.add(optionElement);
+        });
+        y = y + SPACER_PIXELS;
+      } else {
+        // Create text input field for each continuous feature.
+        label.textContent = feature.id;
+        var labelMinMax = designMode.createElement('LABEL', x, y);
+        var min = feature.min.toFixed(2);
+        var max = feature.max.toFixed(2);
+        // Unary plus operator returns a number and truncates trailing zeroes.
+        labelMinMax.textContent = `(min: ${+min}, max: ${+max}):`;
+        labelMinMax.style.width = '300px';
+        y = y + SPACER_PIXELS;
+        var input = designMode.createElement('TEXT_INPUT', x, y);
+        fieldId = alphaNumFeature + '_input';
+        input.id = 'design_' + fieldId;
+        y = y + SPACER_PIXELS;
+      }
+      var addFeature = `data.${alphaNumFeature} = getText("${fieldId}");`;
+      inputFields.push(addFeature);
+    });
+  } else {
```

Comment:
There's a ton of duplication that I'm not going to bother extracting since shortly, I plan to delete this whole branch of the conditional.